Amount units

prusnak opened the pull request:
Fixes https://github.com/trezor/trezor-firmware/issues/324

- [x] add amount units to Protobuf
- [x] implement amount units in Core
- [x] implement amount units in Legacy
- [x] add UI tests

tsusanka commented on core/src/apps/bitcoin/sign_tx/layout.py:
```diff
@@ -15,15 +15,32 @@
+) -> str:
+    decimals, shortcut = coin.decimals, coin.coin_shortcut
+    if amount_unit == AmountUnit.SATOSHI:
+        amount *= 1_0000_0000
```
Quite nit-picky but why isn't styled as `100_000_000`?

tsusanka replied:
Ok, I'm fine with both. Let's leave it then.

tsusanka reviewed (approved):
Core LGTM! Did not review Legacy.

prusnak commented:
There is still one elephant in the room we need to address:

If unit is set to satoshi, we just show `sat` on the display no matter what coin is selected making it impossible to determine which coin is being used. Should we use the `sat BTC`, `sat TEST`, `sat LTC` as suffix? If yes, then we'll probably won't have space to show this on the display :-/

tsusanka commented:
@prusnak how about we leave only "sat" for Bitcoin and "sat [coin]" for all the others? cc @matejzak

prusnak commented:
TT is not an issue:

left = before | right = after

![00000004](https://user-images.githubusercontent.com/42201/105216829-5d42bc80-5b53-11eb-9fb0-b07f35c09a7d.png) ![00000004](https://user-images.githubusercontent.com/42201/105216976-8f541e80-5b53-11eb-957b-d93bad56134e.png)

matejcik commented on common/protob/messages-bitcoin.proto:
```diff
@@ -534,4 +545,5 @@ message AuthorizeCoinJoin {
     repeated uint32 address_n = 4;                                      // prefix of the BIP-32 path leading to the account (m / purpose' / coin_type' / account')
     optional string coin_name = 5 [default='Bitcoin'];                  // coin to use
     optional InputScriptType script_type = 6 [default=SPENDADDRESS];    // used to distinguish between various address formats (non-segwit, segwit, etc.)
+    optional AmountUnit amount_unit = 11;                               // show amounts in
```
let's set `[default=BITCOIN]` so that we don't need to deal with case this is not sent

prusnak replied:
Fix in 3a1b98050463c515587349cf6abbb4090d6ee8ab

matejcik commented on common/protob/messages-bitcoin.proto:
```diff
@@ -170,6 +180,7 @@ message SignTx {
     optional uint32 version_group_id = 8;               // only for Zcash, nVersionGroupId
     optional uint32 timestamp = 9;                      // only for Peercoin
     optional uint32 branch_id = 10;                     // only for Zcash, BRANCH_ID
+    optional AmountUnit amount_unit = 11;               // show amounts in
```
let's set `[default=BITCOIN]` so that we don't need to deal with case this is not sent

prusnak replied:
Fix in 3a1b98050463c515587349cf6abbb4090d6ee8ab

matejcik commented on core/src/apps/bitcoin/sign_tx/approvers.py:
```diff
@@ -42,6 +42,10 @@ def __init__(self, tx: SignTx, coin: CoinInfo) -> None:
         self.orig_total_out = 0  # sum of original output amounts
         self.orig_change_out = 0  # sum of original change output amounts
 
+        self.amount_unit = (
```
let's rely on the default from protobuf

prusnak replied:
Fix in 3a1b98050463c515587349cf6abbb4090d6ee8ab

matejcik commented on core/src/apps/bitcoin/sign_tx/layout.py:
```diff
@@ -15,15 +15,32 @@
+    decimals, shortcut = coin.decimals, coin.coin_shortcut
+    if amount_unit == AmountUnit.SATOSHI:
+        amount *= 1_0000_0000
+        decimals -= 8
```
this seems wrong for coins that don't have 8 decimals?

i'd expect:
```
amount *= 10 ** decimals
decimals = 0
```

matejcik replied:
also the legacy code defends against `decimals < 8` etc, that would probably be a good idea here too

prusnak replied:
Fix in 54085406bbe73202d1c656f26d155de2f80f8bf7

matejcik commented on core/src/apps/bitcoin/authorize_coinjoin.py:
```diff
@@ -61,9 +63,14 @@ async def authorize_coinjoin(ctx: wire.Context, msg: AuthorizeCoinJoin) -> Succe
                 )
             )
         text.normal("Maximum total fees:")
+        amount_unit = (
```
let's rely on the default from protobuf

prusnak replied:
Fix in 3a1b98050463c515587349cf6abbb4090d6ee8ab

matejcik commented on tests/device_tests/test_msg_signtx_amount_unit.py:
```diff
@@ -0,0 +1,71 @@
+
+import pytest
+
+from trezorlib import btc, messages as proto
```
please don't use `as proto` anymore

prusnak replied:
Fix in b0646ecd69fe29ac8ebec7dcc04b6779a06e0781

mcudev commented:
I read a justification for using "bits" and "bitcents" that made a lot of sense to me: https://twitter.com/adam3us/status/1335354727450746881

I understand this can be akin to a religious argument, and I don't want to get into that, but I figured that I'd mention it in the off chance that the terms had not crossed your radar.

prusnak commented:
Found an issue - we should not be multiplying amounts, just decreasing the decimal numbers. Fixed that, but hit some issues with the fixup commits, so I rebased the branch completely on top of the master.
